native/src/main/java/io/ballerina/stdlib/log/MaskedStringBuilder.java

```diff
@@ -0,0 +1,574 @@
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * High-performance builder for creating masked string representations of Ballerina values.
+ * Implements AutoCloseable for proper resource management and memory efficiency.
+ *
+ * @since 2.14.0
+ */
+public class MaskedStringBuilder implements AutoCloseable {
+
+    private static final BString STRATEGY_KEY = StringUtils.fromString("strategy");
+    private static final BString REPLACEMENT_KEY = StringUtils.fromString("replacement");
+    private static final BString EXCLUDE_VALUE = StringUtils.fromString("EXCLUDE");
+    private static final String FIELD_PREFIX = "$field$.";
+    private static final String LOG_ANNOTATION_PREFIX = "ballerina/log";
+    private static final String SENSITIVE_SUFFIX = ":Sensitive";
+
+    private static final BString CYCLIC_REFERENCE_ERROR = StringUtils.fromString("Cyclic value reference detected " +
+            "in the record");
+    public static final BString MASKED_STRING_BUILDER_HAS_BEEN_CLOSED = StringUtils.fromString("MaskedStringBuilder" +
+            " has been closed");
+
+    // Cache for field annotations to avoid repeated extraction
+    private static final Map<RecordType, Map<String, BMap<?, ?>>> ANNOTATION_CACHE = new ConcurrentHashMap<>();
+    private static final int MAX_CACHE_SIZE = 1000;
+
+    // Pre-computed hex lookup table for efficient Unicode escaping
+    private static final char[] HEX_CHARS = "0123456789abcdef".toCharArray();
+
+    private static final char[] QUOTE_ESCAPE = {'\\', '"'};
+    private static final char[] BACKSLASH_ESCAPE = {'\\', '\\'};
+    private static final char[] NEWLINE_ESCAPE = {'\\', 'n'};
+    private static final char[] TAB_ESCAPE = {'\\', 't'};
+    private static final char[] CARRIAGE_RETURN_ESCAPE = {'\\', 'r'};
+    private static final char[] BACKSPACE_ESCAPE = {'\\', 'b'};
+    private static final char[] FORM_FEED_ESCAPE = {'\\', 'f'};
+
+    private final Runtime runtime;
+    private final IdentityHashMap<Object, Boolean> visitedValues;
+    private StringBuilder stringBuilder;
+    private StringBuilder escapeBuffer;
+    private boolean closed = false;
+
+    // Initial capacity configuration
+    private static final int DEFAULT_INITIAL_CAPACITY = 256;
+    private static final int MAX_REUSABLE_CAPACITY = 8192;
+    private static final int ESCAPE_BUFFER_SIZE = 64;
+
+    public MaskedStringBuilder(Runtime runtime) {
+        this.runtime = runtime;
+        this.visitedValues = new IdentityHashMap<>();
+        this.stringBuilder = new StringBuilder(DEFAULT_INITIAL_CAPACITY);
+        this.escapeBuffer = new StringBuilder(ESCAPE_BUFFER_SIZE);
+    }
+
+    public MaskedStringBuilder(Runtime runtime, int initialCapacity) {
+        this.runtime = runtime;
+        this.visitedValues = new IdentityHashMap<>();
+        this.stringBuilder = new StringBuilder(Math.max(initialCapacity, DEFAULT_INITIAL_CAPACITY));
```

Comment:
Is this logic correct? I mean shouldn't we at least want the developer when we are overriding the value they provide?